### DISCUSSION TITLE
BlockingBucket waiting logic

### DISCUSSION THREAD

**Discussion by ShadyD45** [vladimir-bukhtoyarov/bucket4j#321] [Q&A]
Hello @vladimir-bukhtoyarov, Great work on the library! I've some questions about BlockingBucket, Reading the docs I see what it does is it predicts if tokens are going to be available after the specified time, if yes it'll be put in waiting. But if not the request is rejected immediately and I believe this prediction is done based on the refill rate of bucket, please correct me if I'm wrong. 

I've one scenario where I've put a very long refill on bucket but it is rejecting all requests after tokens are exhausted without putting them in waiting and i think my above understanding is the reason. Can you suggest if we can bypass this prediction And just put the request in waiting for given time and then trying to consume then reject if not available?
One workaround I think can work is to have a reasonable refill period and use that refill period to calculate waiting period for each request that comes in after token exhaustion to create an illusion for the prediction logic that tokens will be refilled so that it'll put request in waiting. Can that work?

Thankyou in advance!

**Comment by vladimir-bukhtoyarov** [comment]
Hello @ShadyD45 

> Reading the docs I see what it does is it predicts if tokens are going to be available after the specified time, if yes it'll be put in waiting. But if not the request is rejected immediately and I believe this prediction is done based on the refill rate of bucket, please correct me if I'm wrong.

You are right, blocking bucket behaves in this way. Additionally, bucket proactively decrements amount of tokens in order to protect tokens from to be stolen by parallel threads while current thread is parked. I call this process as reservation. It also need to keep in mind that reservation can temporary make available tokens less than zero.

> I've one scenario where I've put a very long refill on bucket but it is rejecting all requests after tokens are exhausted without putting them in waiting and i think my above understanding is the reason.

1. Try to debug this line https://github.com/bucket4j/bucket4j/blob/1ddc826edfc940b3d1645d707e4cd0d32ce7ecc4/bucket4j-core/src/main/java/io/github/bucket4j/BucketState64BitsInteger.java#L540 or just be sure that you do not request the amount of tokens that greater than bandwidth capacity, it was decided to return false immediately in such case.
2. You also can use the [Verbose API](https://bucket4j.com/8.1.1/toc.html#verbose-api) in order to figure out how many tokens were available at the moment of reject, as well as which capacity was configured for bucket. You can access to VerboseTokenBucket by ```bucket.asBlocking().asVerbose()```;

> One workaround I think can work is to have a reasonable refill period and use that refill period to calculate waiting period for each request that comes in after token exhaustion to create an illusion for the prediction logic that tokens will be refilled so that it'll put request in waiting. Can that work?

To estimate correctness of this workaround I need to have actual numbers: 
* Current bucket configuration.
* Amount of requested tokens that buckets rejects, and second parameter - the amount of waiting time that was passed to try consume.
* The new bucket configuration that is proposed as a fix.
